sql/core/src/test/scala/org/apache/spark/sql/sources/v2/utils/TestV2SessionCatalogBase.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.utils
+
+import java.util
+import java.util.concurrent.ConcurrentHashMap
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.sql.catalog.v2.Identifier
+import org.apache.spark.sql.catalog.v2.expressions.Transform
+import org.apache.spark.sql.execution.datasources.v2.V2SessionCatalog
+import org.apache.spark.sql.sources.v2.Table
+import org.apache.spark.sql.types.StructType
+
+/** A SessionCatalog that always loads an in memory Table, so we can test write code paths. */
+private[v2] trait TestV2SessionCatalogBase[T <: Table] extends V2SessionCatalog {
+
+  protected val tables: util.Map[Identifier, T] = new ConcurrentHashMap[Identifier, T]()
+
+  protected def newTable(
+      name: String,
+      schema: StructType,
+      partitions: Array[Transform],
+      properties: util.Map[String, String]): T
+
+  private def fullIdentifier(ident: Identifier): Identifier = {
+    if (ident.namespace().isEmpty) {
+      Identifier.of(Array("default"), ident.name())
+    } else {
+      ident
+    }
+  }
+
+  override def loadTable(ident: Identifier): Table = {
+    val fullIdent = fullIdentifier(ident)
+    if (tables.containsKey(fullIdent)) {
+      tables.get(fullIdent)
+    } else {
+      // Table was created through the built-in catalog
```

Comment:
I find this case a little odd. I think it makes sense to layer on in-memory tables because we need to return the same table instance. But why create an in-memory shadow table for tables that already exist?

Reply:
Nevermind, I get it.